staticx on 32-bit system

Hi Jonathon,
could you help, is it possible to build staticx on 32-bit system?
I need to bundle a 32-bit executable which successfully runs on obsolete Ubuntu 14.04 and doesn't on 16.04.
Thanks.

```
$ uname -a
Linux ubuntu 3.13.0-170-generic #220-Ubuntu SMP Thu May 9 12:41:17 UTC 2019 i686 i686 i686 GNU/Linux

$ gcc --version | head -n1
gcc (Ubuntu 4.8.4-2ubuntu1~14.04.4) 4.8.4

$ BOOTLOADER_CC=/usr/bin/musl-gcc pip3 install https://github.com/JonathonReinhart/staticx/archive/master.zip
DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.
Collecting https://github.com/JonathonReinhart/staticx/archive/master.zip
  Downloading https://github.com/JonathonReinhart/staticx/archive/master.zip
     - 115 kB 448 kB/s
Collecting pyelftools
  Downloading pyelftools-0.26-py2.py3-none-any.whl (136 kB)
     |████████████████████████████████| 136 kB 367 kB/s 
Building wheels for collected packages: staticx
  Building wheel for staticx (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/pavel/venv/py35/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-up8ipxs5
       cwd: /tmp/pip-req-build-6ko4uyg9/
  Complete output (42 lines):
  running bdist_wheel
  running build
  running build_bootloader
  /usr/bin/musl-gcc -o scons_build/debug/bootloader/error.o -c -std=gnu99 -Wall -Werror -Wmissing-prototypes -Wstrict-prototypes -g -static -DDEBUG=1 -DSTATICX_VERSION=\"0.11.0+g01d4438\" -Ilibtar -Ilibxz bootloader/error.c
  /usr/bin/musl-gcc -o scons_build/debug/bootloader/elfutil.o -c -std=gnu99 -Wall -Werror -Wmissing-prototypes -Wstrict-prototypes -g -static -DDEBUG=1 -DSTATICX_VERSION=\"0.11.0+g01d4438\" -Ilibtar -Ilibxz bootloader/elfutil.c
  /usr/bin/musl-gcc -o scons_build/debug/bootloader/extract.o -c -std=gnu99 -Wall -Werror -Wmissing-prototypes -Wstrict-prototypes -g -static -DDEBUG=1 -DSTATICX_VERSION=\"0.11.0+g01d4438\" -Ilibtar -Ilibxz bootloader/extract.c
  /usr/bin/musl-gcc -o scons_build/debug/bootloader/main.o -c -std=gnu99 -Wall -Werror -Wmissing-prototypes -Wstrict-prototypes -g -static -DDEBUG=1 -DSTATICX_VERSION=\"0.11.0+g01d4438\" -Ilibtar -Ilibxz bootloader/main.c
  bootloader/main.c: In function 'set_rpath':
  bootloader/main.c:163:5: error: format '%lX' expects argument of type 'long unsigned int', but argument 4 has type 'Elf64_Xword' [-Werror=format=]
       debug_printf("Current RPATH (0x%lX): \"%s\"\n", dt_rpath->d_un.d_val,
       ^
  cc1: all warnings being treated as errors
  scons: *** [scons_build/debug/bootloader/main.o] Error 1
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 95, in <module>
      'build':            build_hook,
    File "/home/pavel/venv/py35/lib/python3.5/site-packages/setuptools/__init__.py", line 163, in setup
      return distutils.core.setup(**attrs)
    File "/usr/lib/python3.5/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib/python3.5/distutils/dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/home/pavel/venv/py35/lib/python3.5/site-packages/wheel/bdist_wheel.py", line 290, in run
      self.run_command('build')
    File "/usr/lib/python3.5/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 29, in run
      self.run_command('build_bootloader')
    File "/usr/lib/python3.5/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 24, in run
      check_call(['scons', '-Q'])
    File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['scons', '-Q']' returned non-zero exit status 2
  ----------------------------------------
  ERROR: Failed building wheel for staticx
  Running setup.py clean for staticx
Failed to build staticx
Installing collected packages: pyelftools, staticx
    Running setup.py install for staticx ... error
    ERROR: Command errored out with exit status 1:
     command: /home/pavel/venv/py35/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-4_91bgbk/install-record.txt --single-version-externally-managed --compile --install-headers /home/pavel/venv/py35/include/site/python3.5/staticx
         cwd: /tmp/pip-req-build-6ko4uyg9/
    Complete output (41 lines):
    running install
    running build
    running build_bootloader
    /usr/bin/musl-gcc -o scons_build/debug/bootloader/main.o -c -std=gnu99 -Wall -Werror -Wmissing-prototypes -Wstrict-prototypes -g -static -DDEBUG=1 -DSTATICX_VERSION=\"0.11.0+g01d4438\" -Ilibtar -Ilibxz bootloader/main.c
    bootloader/main.c: In function 'set_rpath':
    bootloader/main.c:163:5: error: format '%lX' expects argument of type 'long unsigned int', but argument 4 has type 'Elf64_Xword' [-Werror=format=]
         debug_printf("Current RPATH (0x%lX): \"%s\"\n", dt_rpath->d_un.d_val,
         ^
    cc1: all warnings being treated as errors
    scons: *** [scons_build/debug/bootloader/main.o] Error 1
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 95, in <module>
        'build':            build_hook,
      File "/home/pavel/venv/py35/lib/python3.5/site-packages/setuptools/__init__.py", line 163, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/pavel/venv/py35/lib/python3.5/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python3.5/distutils/command/install.py", line 583, in run
        self.run_command('build')
      File "/usr/lib/python3.5/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 29, in run
        self.run_command('build_bootloader')
      File "/usr/lib/python3.5/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-6ko4uyg9/setup.py", line 24, in run
        check_call(['scons', '-Q'])
      File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['scons', '-Q']' returned non-zero exit status 2
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/pavel/venv/py35/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-6ko4uyg9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-4_91bgbk/install-record.txt --single-version-externally-managed --compile --install-headers /home/pavel/venv/py35/include/site/python3.5/staticx Check the logs for full command output.
```

--- a/bootloader/elfutil.h
+++ b/bootloader/elfutil.h
@@ -3,11 +3,28 @@
 
 #include <elf.h>
 #include <stdbool.h>
+#include <stdint.h>
 
-#define Elf_Ehdr    Elf64_Ehdr
-#define Elf_Phdr    Elf64_Phdr
-#define Elf_Shdr    Elf64_Shdr
-#define Elf_Dyn     Elf64_Dyn
+/**
+ * Determine native ELF type by looking at size of pointer.
+ * This should work even for x32 which uses elf32:
+ * https://stackoverflow.com/a/36347228
+ */
+#if UINTPTR_MAX == UINT64_MAX
+/* 64-bit */
+# define Elf_Ehdr    Elf64_Ehdr
+# define Elf_Phdr    Elf64_Phdr
+# define Elf_Shdr    Elf64_Shdr
+# define Elf_Dyn     Elf64_Dyn
+#elif UINTPTR_MAX == UINT32_MAX
+/* 32-bit */
+# define Elf_Ehdr    Elf32_Ehdr
+# define Elf_Phdr    Elf32_Phdr
+# define Elf_Shdr    Elf32_Shdr
+# define Elf_Dyn     Elf32_Dyn
+#else
+# error unexpected UINTPTR_MAX
+#endif
 
 bool
 elf_is_valid(const Elf_Ehdr *ehdr);

--- a/bootloader/main.c
+++ b/bootloader/main.c
@@ -2,6 +2,8 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <stdbool.h>
+#include <stdint.h>
+#include <inttypes.h>
 #include <string.h>
 #include <errno.h>
 #include <assert.h>
@@ -160,7 +162,8 @@ dyn_done:
         error(2, 0, "RPATH outside of dynamic strtab!");
     char *rpath = ptr_add(dynstrtab, dt_rpath->d_un.d_val);
 
-    debug_printf("Current RPATH (0x%lX): \"%s\"\n", dt_rpath->d_un.d_val,
+    debug_printf("Current RPATH (0x%"PRIXPTR"): \"%s\"\n",
+            (uintptr_t)dt_rpath->d_un.d_val,
             fmt_str_rep(rpath));
 
     /* Set new RPATH */

